scheduled-messages: Add separate endpoint for editing existing scheduled messages.

laurynmm opened the pull request:
Adds a new separate endpoint for editing scheduled messages: `PATCH /scheduled_messages/<int:scheduled_message_id>`.

Also, removes the `scheduled_message_id` parameter from `POST /scheduled_messages`, and updates that endpoint's `operationId` to be `create-scheduled-message`. Previously, it was `create-or-edit-scheduled-message`.

See [this CZO conversation](https://chat.zulip.org/#narrow/stream/378-api-design/topic/scheduled.20messages/near/1561281) for more context.

---

**Notes**:
- There is a prep commit that adds a second scheduled message to `populate_db.py` for Iago. This is because the `api-test` goes through the `operationId`s in alphabetical order, so the delete curl example happens before the update/edit curl example happens.
- The final three commits can all be squashes before merging:
  - Removes the `scheduled_message_id` parameter from the create scheduled message path.
  - Adds the view and actions for the new edit scheduled message path, with tests.
  - Adds the API documentation for the new endpoint.

**Questions**:
- I made some notes in the code as comments.
- My main question / thought is whether there's a better way to organize the validation aspects of this code path.
  - Currently, if I could do the validation before accessing the scheduled message in question, I did that in the view function. For example, checking if a `type` parameter was sent without a `to` parameter.
  - Then, in the action `edit_scheduled_message`, I checked / validated for information that potentially needed details from the existing scheduled message. For example, if the server had failed to send the scheduled message, in which case we want to require the `scheduled_message_timestamp` to be updated for a time in the future.

---

**Screenshots and screen captures:**

<details>
<summary>API changelog note: feature level 183</summary>


![Screenshot from 2023-05-17 18-39-26](https://github.com/zulip/zulip/assets/63245456/9a192611-e7f8-4715-bbe8-a1d9100df61f)
</details>
<details>
<summary>Edit a scheduled message</summary>

![Screenshot from 2023-05-17 18-48-46](https://github.com/zulip/zulip/assets/63245456/fa576d92-1594-4f20-b310-999294a2fe11)
![Screenshot from 2023-05-17 18-49-02](https://github.com/zulip/zulip/assets/63245456/95adc489-e8d0-4d8b-b8ac-377e3e42d9c9)
![Screenshot from 2023-05-17 18-49-14](https://github.com/zulip/zulip/assets/63245456/fce34e8f-16fa-4277-9ef9-b21da75ac4a4)
</details>
<details>
<summary>Create a scheduled message</summary>

![Screenshot from 2023-05-17 18-50-15](https://github.com/zulip/zulip/assets/63245456/8c4b3a3e-81da-48b9-bde2-69e9c03d7d14)
![Screenshot from 2023-05-17 18-50-28](https://github.com/zulip/zulip/assets/63245456/28fa5535-186a-497c-a81d-93ff0aac953e)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on zerver/actions/scheduled_messages.py:
```diff
@@ -178,7 +251,6 @@ def edit_scheduled_message(
         scheduled_message_object.save()
 
     notify_update_scheduled_message(sender, scheduled_message_object)
-    return scheduled_message_id
```
Noted that the endpoints for editing messages and drafts don't return anything other that the simple success, so updated this accordingly.

laurynmm commented on zerver/tests/test_scheduled_messages.py:
```diff
@@ -433,16 +434,11 @@ def test_editing_failed_send_scheduled_message(self) -> None:
         # successful and resets the `failed` and `failure_message` fields.
```
Once we decide about what to do if the timestamp isn't updated on failed messages, it would be good to add a test for that here.

laurynmm replied:
Note added test coverage here for returning an error in the case that a failed scheduled message is being editing without an updated send timestamp.

laurynmm commented on api_docs/changelog.md:
```diff
@@ -39,7 +48,7 @@ format used by the Zulip server that they are interacting with.
 **Feature level 179**
 
-* [`POST /scheduled_messages`](/api/create-or-update-scheduled-message):
+* [`POST /scheduled_messages`](/api/create-scheduled-message):
```
We perhaps need a redirect for this?

timabbott replied:
I don't think so, it wasn't around for long and I think there aren't any links yet.

laurynmm commented on zerver/openapi/zulip.yaml:
```diff
@@ -18085,6 +18184,25 @@ components:
               "result": "error",
               "stream": "nonexistent_stream",
             }
+    NonExistingStreamIdError:
```
This change is also in a separate PR, #25621, but I included it here to make it an easier rebase process if/when that gets merged.

laurynmm commented on zerver/views/scheduled_messages.py:
```diff
@@ -36,7 +37,76 @@ def delete_scheduled_messages(
+        and message_content is None
+        and scheduled_delivery_timestamp is None
+    ):
+        raise JsonableError(_("Nothing to change"))
```
Added this error for situations where no scheduled message fields are updated. This is similar to a validation check we do in the message edit path.

laurynmm commented on zerver/views/scheduled_messages.py:
```diff
@@ -36,14 +37,82 @@ def delete_scheduled_messages(
+def update_scheduled_message_backend(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    scheduled_message_id: int = REQ(converter=to_non_negative_int, path_only=True),
```
This `converter=to_non_negative_int` is what is used in the `message_edit` view, so went ahead and updated that here as well since this is now in the path. I wasn't sure if we should also make this change for the delete scheduled message path.

timabbott replied:
That's probably a good idea in that context as well, yeah.

laurynmm replied:
#25855 is up with this change.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5265,6 +5254,131 @@ paths:
+            The UNIX timestamp for when the message will be sent,
+            in UTC seconds.
+
+            Required when updating a scheduled message that the server
+            failed to send. See API changelog feature level 181 note for
```
Perhaps:
```suggestion
            Required when updating a scheduled message that the server
            has already tried and failed to send. See API changelog feature level 181 note for
```

so the emphasis is more on the "has already tried" rather than the "failed". I think that makes it easier to see what this requirement is about, because the point is that the old scheduled time is in the past and so we need a new one.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5265,6 +5254,131 @@ paths:
+            in UTC seconds.
+
+            Required when updating a scheduled message that the server
+            failed to send. See API changelog feature level 181 note for
+            more information on identifying scheduled messages that failed.
```
This reference should come with some kind of link (even if words are required to point more specifically to a part of a page that doesn't have its own anchor).

My first reading was actually that this meant some "Changes" note on this page; it was only after not finding one of those for 181 that I realized it meant the changelog at https://zulip.com/api/changelog . We don't currently have links for individual feature levels (though adding those would be a nice touch), but (the relativized version of) https://zulip.com/api/changelog#changes-in-zulip-70 would work well.

---

Separately, I feel like in principle this shouldn't involve referring to any kind of changelog item at all — it's describing the current API, not the differences in any previous API, so it should be self-contained within the description of the current API. If there isn't a good thing to point to in the non-changelog parts of the docs, then that's a sign we should add something.

Here's one version, though:
```suggestion
            failed to send. This state is indicated with `"failed": true` in scheduled-message objects;
            see description at [`GET /scheduled_messages`](/api/get-scheduled-messages).
```
That version leaves implicit that it applies to a "scheduled-message object" as seen in the event system too; alternatively that could be made explicit with some more words.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5265,6 +5254,131 @@ paths:
+          description: |
+            The ID of the scheduled message to update.
+
+            This is different from the unique ID that a message would have
+            after it was sent.
```
Somehow the "was" reads oddly to me here. I think basically I read it as past tense (which then doesn't make sense), and then have to consciously see that it's actually meant as subjunctive.

If you want to keep this in the conditional ("would"), consider evading "was", like:
> would have after being sent

> would have once sent

laurynmm replied:
Yeah, the events descriptions have it in the future subjunctive text which reads much more clearly. I went with this version and updated it for both the edit and delete parameters:

> This is different from the unique ID that a message would have after being sent.

gnprice reviewed (commented):
Thanks! API looks good to me; small comments on docs.

laurynmm commented:
**Updates**:
- Updated for @gnprice's feedback on the API documentation above. See individual comments for specific changes and updated screenshots below.
- Also, added a bit to the `topic` parameter in the `update-scheduled-message` documentation to explicitly state that it's ignored for direct scheduled messages.

**Updated screenshots**:

<details>
<summary>Edit a scheduled message</summary>

*Note that the **Delete a scheduled message** `scheduled_message_id` parameter description has been updated to match the changes here.*
![Screenshot from 2023-05-18 13-36-20](https://github.com/zulip/zulip/assets/63245456/a1002f14-25da-4a3e-96cd-52e92adb3a72)
![Screenshot from 2023-05-18 13-36-49](https://github.com/zulip/zulip/assets/63245456/133b667f-3308-48b3-82af-78f528cf2f7d)
![Screenshot from 2023-05-18 13-37-02](https://github.com/zulip/zulip/assets/63245456/c6a90f5a-3d26-489d-8720-73258a9cca26)
</details>

laurynmm commented:
**Updates**:
- Bumped the API feature level to 184 due to other API change PRs being merged; screenshots haven't been updated.

timabbott commented on zerver/tests/test_home.py:
```diff
@@ -439,7 +439,7 @@ def test_num_queries_for_realm_admin(self) -> None:
         self.login("iago")
         flush_per_request_caches()
-        with self.assert_database_query_count(48):
+        with self.assert_database_query_count(50):
```
I'm a bit puzzled that this added two database queries, not 1; any idea why that is?

(We have these tests to detect database query in loop situations, and I've had kinda thought that the scheduled message stuff only did a single bulk database query to fetch everything).

timabbott commented on zerver/actions/scheduled_messages.py:
```diff
@@ -145,33 +143,111 @@ def notify_update_scheduled_message(
+            # Check message again.
+            addressee = Addressee.legacy_build(
+                sender, updated_recipient_type_name, updated_recipient, updated_topic
+            )
```
I suspect, but am not certain, that this could be extracted into a function that can be shared by the `create_scheduled_message_backend` code that returns an object "Here's the requested content, and the requested recipient type and recipient". I'm not certain that would be cleaner, but it might be.

timabbott commented:
Merged, thanks @laurynmm! I posted a handful of comments on things that might be good follow-up improvements.
